### ISSUE TITLE
aggregation of sparse gradient and dense gradient is unexpected

### ISSUE THREAD

**Issue by albertz** [tensorflow/tensorflow#21243]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOSX 10.13.5
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No
- **TensorFlow installed from (source or binary)**: pip binary
- **TensorFlow version (use command below)**: v1.9.0-0-g25c197e023 1.9.0
- **Python version**: 3.6.5
- **Bazel version (if compiling from source)**: -
- **GCC/Compiler version (if compiling from source)**: -
- **CUDA/cuDNN version**: -
- **GPU model and memory**: -
- **Exact command to reproduce**: see below

### Source code / logs

Example code:

```
    var = tf.get_variable("var", (3, 5), initializer=tf.ones_initializer())
    session.run(var.initializer)
    loss = tf.reduce_sum(tf.nn.embedding_lookup(var, [1]) ** 2) + tf.reduce_sum(var ** 2)
    grad, = tf.gradients(loss, var)
    print("grad:", grad)  # It is an IndexedSlices.
    grad_np = session.run(grad)
    print("grad value:")
    print(grad_np)
```

Output:
```
...
grad value:
IndexedSlicesValue(values=array([[2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2.]], dtype=float32), indices=array([0, 1, 2, 1], dtype=int32), dense_shape=array([3, 5], dtype=int32))
```

The logic in TF is implemented in `_AggregatedGrads` and `_AggregateIndexedSlicesGradients`.
As there is one dense gradient, I would have expected also a final dense gradient.
I think that the dense var updates of the optimizers (e.g. `ApplyAdam`) also perform more efficient compared to the corresponding sparse updates (implemented in `_apply_sparse`).

Regarding how to accumulate the dense and sparse gradients: Probably it can stay like it is, via first accumulating them in `IndexedSlices`, and an additional flag whether there was a dense gradient, and if so, then do a final dense conversion, via `tf.convert_to_tensor`.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 47 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 91 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 122 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@albertz,
Sorry for the delayed response. Upon executing your code on **`Tensorflow Version 1.15.2`**, it is returning **`Dense Gradients`**, not  **`IndexedSlices`**. Please find [the Gist](https://colab.research.google.com/gist/rmothukuru/498380399d69bb013b60ba3c59bf9ef5/gh_21243.ipynb) of the working code and let us know if this is what your expectation is. Thanks!

**Comment by albertz** [comment]
Yes, that is now what I would expect. Thanks!

**Comment by rmothukuru** [comment]
@albertz,
Thank you for the confirmation. Closing the issue as it has been resolved. Thanks!
